tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Grpc/GrpcLegacy/Client/MetadataArraySafeHandleCreateInstrumentation.cs

```diff
@@ -0,0 +1,157 @@
+﻿// <copyright file="MetadataArraySafeHandleCreateInstrumentation.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.ComponentModel;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.DuckTyping;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Grpc.GrpcLegacy.Client
+{
+    /// <summary>
+    /// Grpc.Core.Internal calltarget instrumentation
+    /// </summary>
+    [InstrumentMethod(
+        AssemblyName = "Grpc.Core",
+        TypeName = "Grpc.Core.Internal.MetadataArraySafeHandle",
+        MethodName = "Create",
+        ReturnTypeName = "Grpc.Core.Internal.MetadataArraySafeHandle",
+        ParameterTypeNames = new[] { "Grpc.Core.Metadata" },
+        MinimumVersion = "2.0.0",
+        MaximumVersion = "2.*.*",
+        IntegrationName = nameof(Grpc))]
+    [Browsable(false)]
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    public static class MetadataArraySafeHandleCreateInstrumentation
+    {
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        internal static CallTargetState OnMethodBegin<TInstance, TMetadata>(TInstance instance, TMetadata metadataInstance)
+        {
+            var tracer = Tracer.Instance;
+            if (!GrpcCoreApiVersionHelper.IsSupported || !tracer.Settings.IsIntegrationEnabled(IntegrationId.Grpc))
+            {
+                return CallTargetState.GetDefault();
+            }
+
+            // This integration is called in both the server and client-side paths
+            // and is responsible for serializing the metadata. However, we don't want it it serialize
+            // our Temporary headers, so we remove them before the serialize, and restore them afterwards.
+            // But we only add the extra headers for the client side code, so do a short-circuit check
+            // for one of the temporary headers that should always be there in client side code
+            var metadata = metadataInstance.DuckCast<IMetadata>();
+            if (metadata.Get(TemporaryHeaders.Service) is { } service)
+            {
+                // Remove our temporary headers so they don't get sent over the wire
+                var methodKind = metadata.Get(TemporaryHeaders.MethodKind);
+                var methodName = metadata.Get(TemporaryHeaders.MethodName);
+                var startTime = metadata.Get(TemporaryHeaders.StartTime);
+
+                if (methodKind is not null
```

Comment:
why this if? or put differently, why `metadata.Remove(service)` call below isn't outside this if? It there a risk to send service names on the wire?

Reply:
Hmmm. There's an efficiency vs correctness trade off here. We _always_ add either _all_ the temporary headers, or none of them. This `if` statement is here _primarily_ to stop the null reference checking complaining, but I agree, that's a little bit strange.

> Is there a risk to send service names on the wire?

I don't think there is a _practical_ risk of it, but we could/should make sure. There's no harm in moving the `metadata.Remove()` outside of the if, so I'll do that. In fact, I wonder if we should explicitly do all of them? 🤔 

```csharp
var haveService = metadata.Get(TemporaryHeaders.Service) is { } service
  ? metadata.Remove(service)
  : false;
var haveMethodKind = metadata.Get(TemporaryHeaders.MethodKind) is { } methodKind 
  ? metadata.Remove(methodKind )
  : false;
var haveParentService = metadata.Get(TemporaryHeaders.ParentService) is { } parentService 
  ? metadata.Remove(parentService )
  : false;
// etc
if(haveService && haveMethodKind &&.. )
{
    var temporaryHeaders = new TemporaryGrpcHeaders(
        metadata,
        service: service,
    // ...);
}
```

Reply:
I reworked this in https://github.com/DataDog/dd-trace-dotnet/pull/2545/commits/311d76162052e5291b2146837a8576a597d66a3f to make sure we always remove the headers if we're there. I kept the first `if` so we short-circuit if the expected `service` header isn't there (it won't be there half the time, on the server-side)